[X86][GISel] Add missing legalization for G_IMPLICIT_DEF

bonsthie opened the pull request:
Legalize scalar and vector integer types for `G_IMPLICIT_DEF` at SSE2/AVX2/AVX-512 widths. This is groundwork for upcoming `G_*_VECTOR` legalization, since vector inserts/builds rely on undef bases.

bonsthie committed:
[X86][GISel] Add missing legalization for G_IMPLICIT_DEF

Legalize scalar and vector integer types for G_IMPLICIT_DEF at SSE2/AVX2/AVX-512 widths.
This is groundwork for upcoming G_*_VECTOR legalization, since vector inserts/builds rely on undef bases.

bonsthie committed:
Merge branch 'main' into x86-gisel-legalizer-implicit-def

RKSimon commented on llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:
```diff
@@ -91,7 +91,27 @@ X86LegalizerInfo::X86LegalizerInfo(const X86Subtarget &STI,
+      .legalFor(Is64Bit, {s128})
+      .legalFor(HasSSE2, {v16s8, v8s16, v4s32, v2s64})
+      .legalFor(HasAVX, {v8s32, v4s64})
+      .legalFor(HasAVX2, {v32s8, v16s16, v8s32, v4s64})
```
HasAVX/AVX2 can be merged into a single `.legalFor(HasAVX, {v32s8, v16s16, v8s32, v4s64})` - as all these types are legal on AVX1

Similar for clamping below

bonsthie replied:
I split them because AVX1 lacks 256-bit integer ops, but you’re right that for legalization we only care that the register types exist in the ISA.

RKSimon commented on llvm/lib/Target/X86/GISel/X86LegalizerInfo.cpp:
```diff
@@ -91,7 +91,27 @@ X86LegalizerInfo::X86LegalizerInfo(const X86Subtarget &STI,
+      .legalFor(HasAVX, {v8s32, v4s64})
+      .legalFor(HasAVX2, {v32s8, v16s16, v8s32, v4s64})
+      .legalFor(HasAVX512, {v16s32, v8s64})
+      .legalFor(HasBWI, {v64s8, v32s16})
```
Same as AVX1/2 - HasAVX512 supports all 512-bit vector types

bonsthie committed:
[X86][GISel] Unify AVX/AVX2 and AVX512/AVX512BW type legality

Merge AVX2 types into AVX and AVX512BW types into AVX512, since the vector
types are valid in both cases.

bonsthie committed:
[X86][GISel] Merge builders for G_IMPLICIT_DEF, G_PHI, G_FREEZE, and G_CONSTANT_FOLD_BARRIER

Unify the legality rule sets for value-transport operations (G_IMPLICIT_DEF, G_PHI, G_FREEZE, and G_CONSTANT_FOLD_BARRIER).
These instructions only define/carry values and should therefore accept the same set of legal types.

e-kud commented on llvm/test/CodeGen/X86/GlobalISel/legalize-g_implicit_def.mir:
Nit: we don't have any `g` in test names inside `X86/GlobalISel`. Let's drop it.

e-kud reviewed (commented):
@bonsthie can these newly introduced implicit defs be selected?

bonsthie commented:
> @bonsthie can these newly introduced implicit defs be selected?

You’re right — they all get selected except G_FREEZE. I’ll add a lowering to a copy for `G_FREEZE`, and also add the missing tests.

bonsthie committed:
[X86][GISel] Support G_FREEZE and G_CONSTANT_FOLD_BARRIER selection

- Teach the X86 GlobalISel selector to lower G_FREEZE as a plain COPY, since it
  has no side effects after legalization.
- Add tests to verify selection of G_FREEZE -> COPY and missing coverage for
  G_CONSTANT_FOLD_BARRIER -> COPY.
- Rename legalize-g_implicit_def.mir to legalize-undef-vec-scaling.mir to align
  with existing naming conventions.

RKSimon reviewed (approved):
LGTM

RKSimon committed:
Merge branch 'main' into x86-gisel-legalizer-implicit-def
